funsor/sum_product.py

```diff
@@ -209,17 +213,21 @@ def shift_funsor(f, t):
 
     result = trans(**{time: duration - 1})
     original_names = frozenset(name for name in trans.inputs
-                               if name != time and not name.startswith("P"))
+                               if name != time and name not in global_vars
+                               and not name.startswith("P"))
     for t in range(trans.inputs[time].size - 2, -1, -1):
         result = prod_op(shift_funsor(trans(**{time: t}), duration - t - 1), result)
         sum_vars = frozenset(shift_name(name, duration - t - 1) for name in original_names)
-        result = result.reduce(sum_op, sum_vars)
+        result = result.reduce(sum_op, sum_vars - global_vars)
```

Comment:
nit: should not be necessary since `original_names` already excludes `global_vars`